### DISCUSSION TITLE
Migration to Fabric

### DISCUSSION THREAD

**Discussion by cortinico** [TheWidlarzGroup/react-native-video#4943] [Ideas]
### What problem are you trying to solve?

I'm wondering if someone is looking into migrating React Native Video to use Fabric (i.e. the React Native New Architecture). I'm opening this issue to understand what's the current status of this.

I saw that some of the future work is planned here https://github.com/react-native-video/react-native-video/issues/2642 but nothing is mentioned related to Fabric. I'm wondering if it was intentionally excluded or not.

### Proposed solution

Migrate React Native Video to use Fabric (the React Native New Architecture).

### Additional context

Originally requested here https://github.com/reactwg/react-native-new-architecture/discussions/6#discussioncomment-2584626

**Code sample:** Not a real code sample, but https://github.com/react-native-community/RNNewArchitectureLibraries contains a sample of a library that is backward compatible with both the old and the new arch of React Native.

EDIT: typo

**Comment by douglowder** [comment]
I'm willing to take a look at this since I maintain the RN TV repo, and video is kind of important for TV :)

**Comment by hueniverse** [comment]
It's on my list but the top priority is to get to zero pending PRs and less than 100 open issues.

**Comment by mk0116** [comment]
I can help you @hueniverse. Let me know if you need any help from me to migrate to new arch.

**Comment by hueniverse** [comment]
@mk0116 thanks for the offer! Let's get v6 out first and when things are a bit more stable, we can move forward with it.

**Comment by capezzbr** [comment]
Hey @mk0116 @hueniverse, any update on supporting the new architecture?

**Comment by hueniverse** [comment]
@capezzbr unless someone shows up and does the work, I don't see it happening in the near future. So far no one has expressed interest in taking a lead on this.

**Comment by mk0116** [comment]
I'm waiting for v6 to roll out. It seems like v6 has a lot of changes with breaking changes. So that after rolling out v6, I guess that's a better timing we can aim to new arch migration.

**Comment by cipolleschi** [comment]
Hey there! @mk0116 is there any estimated date for the v6 to be released?

**Comment by mk0116** [comment]
I'm not a maintainer of this library @cipolleschi I just suggested to help them migrate to new arch and @hueniverse said rolling out v6 is the priority right now.

**Comment by hari-mohan-choudhary** [comment]
I am using v6.0.0-alpha.4 but this is not support fabric. any way I can use in fabric

**Comment by samjayhk** [comment]
@hari-mohan-choudhary Hello, did you dose any additional step for using `react-native-video` under Fabric mode? I have tried `v6.0.0-alpha.4` and `v6.0.0-alpha.5`, but it still showing `Unimplemented Components <Video>`. Thanks.

**Comment by hari-mohan-choudhary** [comment]
@samjayhk I am getting the same error.

**Comment by prashantchothani** [comment]
@douglowder request you to please let me know if the new architecture support for tvos react native for react native video is done ?  Referring to below.

> I'm willing to take a look at this since I maintain the RN TV repo, and video is kind of important for TV :)

**Comment by freeboub** [comment]
New architecture is not yet considered as stable, and nobody open a PR to support it. So RNV doesn't support it yet. But I agree this is a must have in the future!

**Comment by douglowder** [comment]
@prashantchothani I have not yet picked up this work

**Comment by Sunbreak** [comment]
Working on fabric support: https://github.com/react-native-video/react-native-video/pull/3056

**Comment by BraveEvidence** [comment]
This will help https://www.youtube.com/watch?v=vVI7ZAZq5e0

**Comment by jiyong1** [comment]
could you please take a look at this [PR](https://github.com/react-native-video/react-native-video/pull/3122)

**Comment by sneh2608** [comment]
> could you please take a look at this [PR](https://github.com/react-native-video/react-native-video/pull/3122)
@jiyong1 
I have used your PR in my Project...It seems that there is some bug in the implementation.
Even when the video component is unmounted the video plays in background even when playInBackground flag is passed as false.I am currently implementing a carousel which has video,lottie and image support.
Tried various method but the video instance won't get destroyed
Could you please look into this?
```
<Video
          repeat={false}
          ref={videoRef}
          resizeMode="contain"
          paused={pause || loading}
          volume={1}
          source={{uri: convertToProxyURL(url)}}
          onEnd={onVideoEnd}
          onError={(_error: any) => {
            setLoading(false);
          }}
          onProgress={data => {
            if (isCurrentIndex) {
              onVideoProgress?.(data);
            }
          }}
          bufferConfig={{
            minBufferMs: BUFFER_TIME,
            maxBufferMs: BUFFER_TIME,
            bufferForPlaybackMs: BUFFER_TIME,
            bufferForPlaybackAfterRebufferMs: BUFFER_TIME,
          }}
          onBuffer={onBuffer}
          onLoadStart={onLoadStart}
          onLoad={(item: OnLoadData) => {
            videoData.current = item;
            loadVideo();
          }}
          style={styles.contentVideoView}
          playInBackground={false}
        />
```
Edit  - I was able to solve this issue by implementing the onDropViewInstance method in ReactExoplayerViewmanager file and clearing the view resources when the component unmounts
`@Override
    public void onDropViewInstance(ReactExoplayerView view) {
        view.cleanUpResources();
        ((ThemedReactContext) view.getContext()).removeLifecycleEventListener(view);
    }`

**Comment by AbdouEB** [comment]
Hey, any updates on fabric support on Android as it seems like it is still an issue..

**Comment by freeboub** [comment]
could you please take a look at this https://github.com/react-native-video/react-native-video/pull/3122

**Comment by jessiemblack** [comment]
Having this issue still with RN 0.72.6. Not able to use the library

**Comment by theyanniss23002** [comment]
How long will it take before you get around to developing an update to support fabric?

**Comment by cipolleschi** [comment]
Hello there! Have you tried to run the `react-native-video` component through the [interop layers](https://github.com/reactwg/react-native-new-architecture/discussions/135) we provided starting from 0.72? This should be the way to use the component in a New Architecture app while it is still on the old architecture.

Also, if you tried that path, and there are issues with `react-native-video`, we would love to look at them, as we might be able to address them in the interop layer and that would help also other libraries/components to work better in the New Architecture through the interop layers.

**Comment by theyanniss23002** [comment]
@cipolleschi I'm using react-native version 0.73, is it ok?

**Comment by cipolleschi** [comment]
sure, even better. We improved the Interop Layers in 0.73!

**Comment by freeboub** [comment]
> How long will it take before you get around to developing an update to support fabric?

Nobody can provide an eta, It would be faster with more reviewers / contributors 😅
That said there are some pull requests open to integrate it!

**Comment by theyanniss23002** [comment]
> sure, even better. We improved the Interop Layers in 0.73!

Then I ran into a problem... version 0.73, maybe something needs to be changed in this case, but so far I have no idea what exactly
<img width="489" alt="Снимок экрана 2024-01-25 в 10 28 06" src="https://github.com/react-native-video/react-native-video/assets/57316921/3607c6fa-8ca5-4e61-9136-ed05902bf843">
<img width="330" alt="Снимок экрана 2024-01-25 в 10 28 17" src="https://github.com/react-native-video/react-native-video/assets/57316921/5d10ea82-4783-4d22-885c-ec2d3f7c2f6f">
<img width="322" alt="Снимок экрана 2024-01-25 в 10 28 32" src="https://github.com/react-native-video/react-native-video/assets/57316921/403aa9fe-bcf4-4bf9-a510-5997f7b70540">

**Comment by cipolleschi** [comment]
@theyanniss23002 have you registered your component in the `react-native.config.js` as explained in the [`How to use the interop layer` section](https://github.com/reactwg/react-native-new-architecture/discussions/135)?

> ## How to use the interop layer
> The interop layer is shipped with React Native 0.72. If no components are registered, it does nothing.
The only step required to use it is to register a component in the layer.
>
> To register a component, you have to edit the react-native.config.js file in your application project to list the legacy components required by the app, in the Android and iOS sections:
>```
>module.exports = {
>    project:{
>      android: {
>        unstable_reactLegacyComponentNames: [ "ComponentName" ]
>      },
>      ios: {
>        unstable_reactLegacyComponentNames: [ "ComponentName" ]
>      }
>    },
>};
>```

**Comment by theyanniss23002** [comment]
> @theyanniss23002 have you registered your component in the `react-native.config.js` as explained in the [`How to use the interop layer` section](https://github.com/reactwg/react-native-new-architecture/discussions/135)?
> 
> > ## How to use the interop layer
> > The interop layer is shipped with React Native 0.72. If no components are registered, it does nothing.
> > The only step required to use it is to register a component in the layer.
> > To register a component, you have to edit the react-native.config.js file in your application project to list the legacy components required by the app, in the Android and iOS sections:
> > ```
> > module.exports = {
> >    project:{
> >      android: {
> >        unstable_reactLegacyComponentNames: [ "ComponentName" ]
> >      },
> >      ios: {
> >        unstable_reactLegacyComponentNames: [ "ComponentName" ]
> >      }
> >    },
> > };
> > ```

Thank you! I did it

**Comment by Mahmoud412** [comment]
@cipolleschi i have been trying to use interop layer for react native video 6.0.0-alpha.1 and i follow How to use the interop layer section but still not working.
i'm using React Native 0.72
**react-native.config.js**
```
module.exports = {
   project:{
     android: {
       unstable_reactLegacyComponentNames: [ "RCTVideo" ]
     },
     ios: {
       unstable_reactLegacyComponentNames: [ "" ]
     }
   },
};
```

**Comment by theyanniss23002** [comment]
> @cipolleschi i have been trying to use interop layer for react native video 6.0.0-alpha.1 and i follow How to use the interop layer section but still not working. i'm using React Native 0.72 **react-native.config.js**
> 
> ```
> module.exports = {
>    project:{
>      android: {
>        unstable_reactLegacyComponentNames: [ "RCTVideo" ]
>      },
>      ios: {
>        unstable_reactLegacyComponentNames: [ "" ]
>      }
>    },
> };
> ```

Use this config

```js
module.exports = {
  project: {
    android: {
      unstable_reactLegacyComponentNames: ['Video'],
    },
    ios: {
      unstable_reactLegacyComponentNames: ['Video'],
    },
  },
};
```

**Comment by Mahmoud412** [comment]
@theyanniss23002 now i'm getting this error and the app crashing right after it. 
<img width="307" alt="android" src="https://github.com/react-native-video/react-native-video/assets/52178976/ad0675d5-a6f8-4a56-bf70-e19e453ec8e2">

**Comment by theyanniss23002** [comment]
> @theyanniss23002 now i'm getting this error and the app crashing right after it. <img alt="android" width="307" src="https://private-user-images.githubusercontent.com/52178976/304000982-ad0675d5-a6f8-4a56-bf70-e19e453ec8e2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc3MjkwMzgsIm5iZiI6MTcwNzcyODczOCwicGF0aCI6Ii81MjE3ODk3Ni8zMDQwMDA5ODItYWQwNjc1ZDUtYTZmOC00YTU2LWJmNzAtZTE5ZTQ1M2VjOGUyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjEyVDA5MDUzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMzZGNiZWQ5NTI1OGQ2Nzg1YzUwMzIyMjQ0YTcxOTAyOWE0MzUzMWQ1YzAxZDNmNzE4NjAxM2VmOTFiODA5NmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.kRtbnyOYBXhIqGj1SRUfRhtZgaXuVELgo2Kp4iI8ibw">

I can't imagine why you are getting this error. contact the library creators

**Comment by KrzysztofMoch** [comment]
@Mahmoud412 You need to update to 6.0.0-beta.5 (it includes fix for that)

**Comment by TWG-AI** [comment]
Good news — v7 of react-native-video is already built entirely on the New Architecture via React Native Nitro Modules and Fabric (`codegenNativeComponent`). If new-arch support is what you need, upgrading to v7 is the path forward.

> I'm an AI assistant — worth double-checking the migration guide in the v7 docs before upgrading.
